persist/encoding/stream.go

```diff
@@ -0,0 +1,148 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE
+
+package encoding
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+)
+
+// DecoderStream is a data stream that is read by the decoder,
+// it takes both a reader and the underlying backing bytes.
+// This is constructed this way since the decoder needs access
+// to the backing bytes when taking refs directly for decoding byte
+// slices without allocating bytes itself but also needs to progress
+// the reader (for instance when a reader is a ReaderWithDigest that
+// is calculating a digest as its being read).
+type DecoderStream interface {
+	io.Reader
+
+	// Reset resets the decoder stream for decoding a new byte slice.
+	Reset(b []byte)
+
+	// ReadByte reads the next byte.
+	ReadByte() (byte, error)
+
+	// UnreadByte unreads the last read byte or returns error if none read
+	// yet. Only a single byte can be unread at a time, a consecutive call
+	// to UnreadByte will result in an error.
+	UnreadByte() error
+
+	// Bytes returns the ref to the backing bytes, note decoders will
+	// call Skip if they "read" any bytes by simply taking refs to them
+	// so it is guaranteed that Read/Skip is called until EOF is reached.
+	Bytes() []byte
+
+	// Skip progresses the reader by a certain amount of bytes, useful
+	// when taking a ref to some of the bytes and progressing the reader
+	// itself.
+	Skip(length int64) error
+
+	// Remaining returns the remaining bytes in the stream.
+	Remaining() int64
+}
+
+type decoderStream struct {
+	reader       *bytes.Reader
+	bytes        []byte
+	lastReadByte int
+	unreadByte   int
+}
+
+// NewDecoderStream creates a new decoder stream from a bytes ref.
+func NewDecoderStream(b []byte) DecoderStream {
+	return &decoderStream{
+		reader:       bytes.NewReader(b),
+		bytes:        b,
+		lastReadByte: -1,
+		unreadByte:   -1,
+	}
+}
+
+func (s *decoderStream) Reset(b []byte) {
+	s.reader.Reset(b)
+	s.bytes = b
+	s.lastReadByte = -1
+	s.unreadByte = -1
+}
+
+func (s *decoderStream) Read(p []byte) (int, error) {
+	ref := p
+
+	var numUnreadByte int
+	if s.unreadByte >= 0 {
+		p[0] = byte(s.unreadByte)
```

Comment:
do you need to guarantee that `len(p)` > 0 here?

Reply:
Yes, I'll add what I was going to originally do with a `if len(p) == 0 { return 0, nil }` at the beginning of the method (but was upset that I'd have to add a test for branch coverage ha).